library/ssl_srv.c

```diff
@@ -2907,55 +2907,86 @@ static int ssl_get_ecdh_params_from_cert( mbedtls_ssl_context *ssl )
     psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
     unsigned char buf[
         PSA_KEY_EXPORT_ECC_KEY_PAIR_MAX_SIZE(PSA_VENDOR_ECC_MAX_CURVE_BITS)];
-    psa_key_attributes_t key_attributes;
+    psa_key_attributes_t key_attributes = PSA_KEY_ATTRIBUTES_INIT;
     size_t ecdh_bits = 0;
     size_t key_len;
+    mbedtls_pk_context *pk;
+    mbedtls_ecp_keypair *key;
 
-    if( ! mbedtls_pk_can_do( mbedtls_ssl_own_key( ssl ), MBEDTLS_PK_ECKEY ) )
-    {
-        return( MBEDTLS_ERR_SSL_PK_TYPE_MISMATCH );
-    }
+    pk = mbedtls_ssl_own_key( ssl );
 
-    mbedtls_ecp_keypair *key =
-            mbedtls_pk_ec( *mbedtls_ssl_own_key( ssl ) );
-
-    if( key == NULL )
+    if( pk == NULL )
         return( MBEDTLS_ERR_ECP_BAD_INPUT_DATA );
 
-    /* Convert EC group to PSA key type. */
-    if( ( ssl->handshake->ecdh_psa_type =
-                mbedtls_ecc_group_to_psa( key->grp.id,
-                                          &ecdh_bits ) ) == 0 )
+    switch( mbedtls_pk_get_type( pk ) )
     {
-        return( MBEDTLS_ERR_SSL_HANDSHAKE_FAILURE );
-    }
+    case MBEDTLS_PK_OPAQUE:
+        if( ! mbedtls_pk_can_do( pk, MBEDTLS_PK_ECKEY ) )
+            return( MBEDTLS_ERR_SSL_PK_TYPE_MISMATCH );
 
-    if( ecdh_bits > 0xffff )
-        return( MBEDTLS_ERR_SSL_ILLEGAL_PARAMETER );
+        ssl->handshake->ecdh_psa_privkey =
+                *( (mbedtls_svc_key_id_t*) pk->pk_ctx );
 
-    ssl->handshake->ecdh_bits = (uint16_t) ecdh_bits;
+        status = psa_get_key_attributes( ssl->handshake->ecdh_psa_privkey,
+                                         &key_attributes );
+        if( status != PSA_SUCCESS)
+            return( MBEDTLS_ERR_PK_BAD_INPUT_DATA );
```

Comment:
Is there a specific reason to return that code rather than use `psa_ssl_status_to_mbedtls()`? If yes, please add a comment; otherwise please use `psa_ssl_status_to_mbedtls()`.